File: src/Weekly_Contest_198/Number_of_Nodes_in_the_Sub_Tree_With_the_Same_Label/Solution.kt
package Weekly_Contest_198.Number_of_Nodes_in_the_Sub_Tree_With_the_Same_Label

fun main() {
    Solution().countSubTrees(
        2, arrayOf(
            intArrayOf(0, 1)
        ), "ab"
    )
}

// https://leetcode.com/contest/weekly-contest-198/problems/number-of-nodes-in-the-sub-tree-with-the-same-label/
class Solution {
    class Node(
        val i: Int,
        val char: Char,
        var children: HashSet<Node> = hashSetOf(),
        val map: IntArray = IntArray(26)
    )

    fun addArray(a: IntArray, b: IntArray) {
        repeat(26) { i ->
            a[i] += b[i]
        }
    }

    fun countSubTrees(n: Int, edges: Array<IntArray>, labels: String): IntArray {
        val nodes = Array(n) { i ->
            Node(i, labels[i])
        }

        edges.forEach { edge ->
            nodes[edge[0]].children.add(nodes[edge[1]])
            nodes[edge[1]].children.add(nodes[edge[0]])
        }

        nodes[0].step()

        return nodes.map { it.map[it.char.toInt()-97] }.toIntArray()
    }

    fun Node.step(parent: Node? = null) {
        if (parent != null) children.remove(parent)
        this.map[this.char.toInt() - 97] = 1
        children.forEach { child ->
            child.step(this)
            addArray(this.map, child.map)
        }
    }
}

File: src/Weekly_Contest_198/Number_of_Nodes_in_the_Sub_Tree_With_the_Same_Label/Solution.py
from typing import List, Optional


class Node:
    def __init__(self, i: int, char: str):
        self.i = i
        self.char = char
        self.children = set()
        self.map = [0] * 26  # index(character) -> count in sub tree


class Solution:
    def countSubTrees(self, n: int, edges: List[List[int]], labels: str) -> List[int]:
        nodes = [Node(i, labels[i]) for i in range(n)]
        for edge in edges:
            nodes[edge[0]].children.add(nodes[edge[1]])
            nodes[edge[1]].children.add(nodes[edge[0]])

        def add_array(a: List[int], b: List[int]):
            for i in range(26):
                a[i] += b[i]

        def step(node: Node, parent: Optional[Node]):
            if parent:
                node.children.remove(parent)
            node.map[ord(node.char) - 97] = 1
            for child in node.children:
                step(child, node)
                add_array(node.map, child.map)

        step(nodes[0], None)

        return list(map(lambda node: node.map[ord(node.char) - 97], nodes))
